src/libp2p_crypto.erl

```diff
@@ -296,14 +346,94 @@ from_network(testnet) -> ?NETTYPE_TEST.
 to_network(?NETTYPE_MAIN) -> mainnet;
 to_network(?NETTYPE_TEST) -> testnet.
 
+-spec key_size_bytes(non_neg_integer()) -> pos_integer().
+key_size_bytes(?KEYTYPE_ED25519) ->
+    32;
+key_size_bytes(?KEYTYPE_ECC_COMPACT) ->
+    32;
+key_size_bytes(KeyType) ->
+    error({bad_key_type, KeyType}).
+
 %% @doc Verifies a binary against a given digital signature over the
 %% sha256 of the binary.
 -spec verify(binary(), binary(), pubkey()) -> boolean().
+verify(Bin, MultiSignature, {multisig, M, N, KeysDigest}) ->
+    try
+        {Keys, KeysLen} = multisig_parse_keys(MultiSignature, N),
+        N = length(Keys),
+        <<KeysBin:KeysLen/binary, ISigsBin/binary>> = MultiSignature,
+        case multihash:hash(KeysDigest) of
+            {error, _} ->
+                false;
+            {ok, HashType} ->
+                case multihash:digest(KeysBin, HashType) of
+                    {error, _} ->
+                        false;
+                    {ok, KeysDigest} ->
+                        ISigs = multisig_parse_isigs(ISigsBin, M, N),
+                        %% Reject dup key index
+                        case ISigs -- lists:ukeysort(1, ISigs) of
+                            [] ->
+                                %% Index range: 0..N-1
+                                KS = [{lists:nth(I + 1, Keys), S} || {I, S} <- ISigs],
+                                M =< length([{} || {K, S} <- KS, verify(Bin, S, K)]);
+                            [_|_] ->
```

Comment:
would _ not suffice here?

Reply:
It's a general pattern I forced myself to make a habit of - match as much structure as possible, for 2 benefits:
1. effectively assertions to keep unpredictable mistakes from propagating
2. explicit expressions help a code reader's confidence

Reply:
fair enough